daemon/graphdriver/zfs/zfs.go

```diff
@@ -0,0 +1,277 @@
+		return graphdriver.ErrPrerequisites
+	}
+	return nil
+}
+
+var CprocMounts = C.CString("/proc/mounts")
+var CopenMod = C.CString("r")
+
+func lookupZfsDataset(rootdir string) (string, error) {
+	var stat syscall.Stat_t
+	var Cmnt C.struct_mntent
+	var Cfp *C.FILE
+	buf := string(make([]byte, 256, 256))
+	Cbuf := C.CString(buf)
+	defer free(Cbuf)
+
+	if err := syscall.Stat(rootdir, &stat); err != nil {
+		return "", fmt.Errorf("Failed to access '%s': %s", rootdir, err)
+	}
+	wantedDev := stat.Dev
+
+	if Cfp = C.setmntent(CprocMounts, CopenMod); Cfp == nil {
+		return "", fmt.Errorf("Failed to open /proc/mounts")
+	}
+	defer C.endmntent(Cfp)
+
+	for C.getmntent_r(Cfp, &Cmnt, Cbuf, 256) != nil {
+		dir := C.GoString(Cmnt.mnt_dir)
+		if err := syscall.Stat(dir, &stat); err != nil {
+			log.Debugf("[zfs] failed to stat '%s' while scanning for zfs mount: %v", dir, err)
+			continue // may fail on fuse file systems
+		}
+
+		fs := C.GoString(Cmnt.mnt_type)
+		if stat.Dev == wantedDev && fs == "zfs" {
+			return C.GoString(Cmnt.mnt_fsname), nil
+		}
+	}
+	// should never happen
+	return "", fmt.Errorf("Failed to find zfs pool in /proc/mounts")
+}
+
+func free(p *C.char) {
+	C.free(unsafe.Pointer(p))
+}
+
+type Driver struct {
+	dataset *zfs.Dataset
+	options ZfsOptions
+}
+
+func (d *Driver) String() string {
+	return "zfs"
+}
+
+func (d *Driver) Cleanup() error {
+	return nil
+}
+
+func (d *Driver) Status() [][2]string {
+	parts := strings.Split(d.dataset.Name, "/")
+	pool, err := zfs.GetZpool(parts[0])
+
+	if err != nil {
+		return [][2]string{
+			{"error while getting pool", fmt.Sprintf("%v", err)},
+		}
+	}
+	var quota string
+	if d.dataset.Quota == 0 {
+		quota = strconv.FormatUint(d.dataset.Quota, 10)
+	} else {
+		quota = "no"
+	}
+
+	return [][2]string{
+		{"Zpool", pool.Name},
+		{"Zpool Health", pool.Health},
+		{"Parent Dataset", d.dataset.Name},
+		{"Space Used By Parent", strconv.FormatUint(d.dataset.Used, 10)},
+		{"Space Available", strconv.FormatUint(d.dataset.Avail, 10)},
+		{"Parent Quota", quota},
+		{"Compression", d.dataset.Compression},
+	}
+}
+
+func cloneFilesystem(id, parent, mountpoint string) error {
+	parentDataset, err := zfs.GetDataset(parent)
+	if parentDataset == nil {
+		return err
+	}
+	snapshotName := fmt.Sprintf("%d", time.Now().Nanosecond())
+	snapshot, err := parentDataset.Snapshot(snapshotName /*recursive */, false)
+	if snapshot == nil {
+		return err
+	}
+
+	_, err = snapshot.Clone(id, map[string]string{
+		"mountpoint": mountpoint,
+	})
+	if err != nil {
+		snapshot.Destroy(zfs.DestroyDeferDeletion)
+		return err
+	}
+	err = snapshot.Destroy(zfs.DestroyDeferDeletion)
+	return err
+}
+
+func (d *Driver) ZfsPath(id string) string {
+	return d.options.fsName + "/" + id
+}
+
+func (d *Driver) Create(id string, parent string) error {
+	mountPoint := path.Join(d.options.mountPath, "graph", id)
+	datasetName := d.ZfsPath(id)
+	dataset, err := zfs.GetDataset(datasetName)
```

Comment:
if `dataset.Destroy(zfs.DestroyRecursiveClones)` returns an error, there are two options:
- if dataset was already destroyed (race condition between GetDataset and Destroy)
- it failed to destroying the dataset, in this case zfs.CreateFilesystem will also fail
I will use log.Warnf instead